src/main/java/org/gridsuite/odre/server/utils/GeographicDataParser.java

```diff
@@ -82,7 +82,34 @@ private GeographicDataParser() {
         return substations;
     }
 
-    public static Map<String, LineGeoData> parseLines(BufferedReader aerialLinesBr, BufferedReader undergroundLinesBr) {
+    private static double distanceCoordinate(Coordinate coord1, Coordinate coord2) {
+        return DistanceCalculator.distance(coord1.getLat(), coord1.getLon(), coord2.getLat(), coord2.getLon());
+    }
+
+    public static String findSubstationStart(Map<String, SubstationGeoData> substationGeoData, String lineId, List<Coordinate> coordinates) {
+        String substation1 = lineId.substring(0, 5).trim();
+        String substation2 = lineId.substring(8).trim();
+        SubstationGeoData geo1 = substationGeoData.get(substation1);
+        SubstationGeoData geo2 = substationGeoData.get(substation2);
+
+        if (geo1 == null && geo2 == null) {
+            return "";
+        } else if (geo1 != null && geo2 != null) {
+            return distanceCoordinate(geo1.getCoordinate(), coordinates.get(0))
+                    < distanceCoordinate(geo2.getCoordinate(), coordinates.get(0))
+                    ? substation1 : substation2;
+        } else {
+            double d1 = distanceCoordinate((geo1 != null ? geo1 : geo2).getCoordinate(), coordinates.get(0));
+            double d2 = distanceCoordinate((geo1 != null ? geo1 : geo2).getCoordinate(), coordinates.get(coordinates.size() - 1));
+            if (d1 > d2) {
+                Collections.reverse(coordinates);
```

Comment:
why reverse coordinates in this case but not in the previous case ?

Reply:
we know the substation id, it's more efficient this way, in se second case, we are missing the substation id near the start, so we reverse the line to get the substationStart corresponding to the line

Reply:
So it's because you store only the substationstart ? If we store substationStart and substationEnd we don't have to reverse this anymore ?

Reply:
if we found only the second one, we will need to reverse it anyway

Reply:
why ?